clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/GoogleExecutorTraits.groovy

```diff
@@ -0,0 +1,86 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.clouddriver.google
+
+import com.google.api.client.googleapis.batch.BatchRequest
+import com.google.api.client.googleapis.services.AbstractGoogleClientRequest;
+import com.netflix.spectator.api.Clock
+import com.netflix.spectator.api.Id
+import com.netflix.spectator.api.Registry
+
+import java.io.IOException
+import java.util.concurrent.TimeUnit;
+
+
+trait GoogleExecutorTraits {
+  final String TAG_BATCH_CONTEXT = "context"
+  final String TAG_REGION = "region"
+  final String TAG_SCOPE = "scope"
+  final String TAG_ZONE = "zone"
+  final String SCOPE_BATCH = "batch"
+  final String SCOPE_GLOBAL = "global"
+  final String SCOPE_REGIONAL = "regional"
+  final String SCOPE_ZONAL = "zonal"
+
+  abstract Registry getRegistry()
+
+  public void timeExecuteBatch(BatchRequest batch, String batchContext, String... tags) throws IOException {
+     Registry registry = getRegistry()
+     Clock clock = registry.clock()
+     Id id = registry.createId("google.batch.execute", tags).withTags(TAG_BATCH_CONTEXT, batchContext, TAG_SCOPE, SCOPE_BATCH)
+     Id sizeId = registry.createId("google.batchSize", tags).withTag(TAG_BATCH_CONTEXT, batchContext)
+
+     long startTime = clock.monotonicTime()
+
+     try {
+       batch.execute()
+       long nanos = clock.monotonicTime() - startTime
+       registry.timer(id.withTag("success", "true")).record(nanos, TimeUnit.NANOSECONDS)
+       registry.counter(sizeId.withTag("success", "true")).increment(batch.size())
+     } catch (IOException ioex) {
+       registry.timer(id.withTag("success", "false")).record(clock.monotonicTime() - startTime, TimeUnit.NANOSECONDS)
+       registry.counter(sizeId.withTag("success", "false")).increment(batch.size())
+       throw ioex
+     } catch (Exception ex) {
+       registry.timer(id.withTag("success", "false")).record(clock.monotonicTime() - startTime, TimeUnit.NANOSECONDS)
+       registry.counter(sizeId.withTag("success", "false")).increment(batch.size())
+       throw new IllegalStateException(ex);
+     }
+  }
+
+  public <T> T timeExecute(AbstractGoogleClientRequest<T> request, String api, String... tags) throws IOException {
+     Registry registry = getRegistry()
+     Clock clock = registry.clock()
+     Id id = registry.createId("google.api", tags).withTag("api", api)
+     long startTime = clock.monotonicTime();
+
+     try {
+       T result = request.execute()
+       long nanos = clock.monotonicTime() - startTime
+       registry.timer(id.withTag("success", "true")).record(nanos, TimeUnit.NANOSECONDS)
+       return result
+     } catch (IOException ioex) {
+       registry.timer(id.withTag("success", "false")).record(clock.monotonicTime() - startTime, TimeUnit.NANOSECONDS)
+       throw ioex;
+     } catch (Exception ex) {
```

Comment:
The IOException is handled different from the others. It is passed through. The caller might grab the HTTP response from HTTP-like exceptions. My thought here was that API calls are going to fail IO or HTTP exceptions (which are derived from IO) and get reported back. All the other exceptions and errors are more likely internal to Spinnaker implementation/configuration rather than user requests, resource state, and permissions.